Export spectral subsets

rosteen opened the pull request:
Implements updating spectral subsets via `specutils.SpectralRegion` to ECSV files.

rosteen commented:
Hmm, this was passing locally.

pllim commented on jdaviz/configs/default/plugins/export/tests/test_export.py:
```diff
@@ -205,6 +205,18 @@ def test_basic_export_subsets_cubeviz(self, cubeviz_helper, spectral_cube_wcs):
+
+        export_plugin.filename_value = "test_spectral_region"
+        export_plugin.export()
+        assert os.path.isfile('test_spectral_region.ecsv')
```
Is roundtripping a requirement? That is, is Jdaviz expected to be able to re-ingest this file and re-creates the same spectral region?

rosteen replied:
I don't believe it is.

pllim replied:
OK. I guess checking for the output content is correct or not is the job of `specutils`, so we don't have to do that here. Thanks for the clarification!

pllim commented on pyproject.toml:
```diff
@@ -23,7 +23,7 @@ dependencies = [
     "voila>=0.4,<0.5",
     "pyyaml>=5.4.1",
-    "specutils>=1.9",
+    "specutils>=1.15",
```
That is quite a jump. Do we need a change log for this?

rosteen replied:
I would be happy to add it to the change log if you think so.

pllim replied:
Doesn't hurt?

rosteen commented on jdaviz/configs/default/plugins/export/tests/test_export.py:
```diff
@@ -93,10 +93,6 @@ def test_not_implemented(self, cubeviz_helper, spectral_cube_wcs):
 
-        cubeviz_helper.app.session.edit_subset_mode.mode = NewMode
-        cubeviz_helper.app.get_viewer("spectrum-viewer").apply_roi(XRangeROI(5, 15.5))
-        assert 'Subset 2' not in export_plugin.subset.choices
```
It means Subset 2 is allowed to be in the subset choices now.

pllim replied:
But is it grayed out? If so, is there a way to check?

rosteen replied:
It is not grayed out, no. The output format choices may be, I can add a check for that.

pllim commented on jdaviz/configs/default/plugins/export/export.py:
```diff
@@ -655,6 +696,10 @@ def save_subset_as_region(self, selected_subset_label, filename):
 
         region.write(filename, overwrite=True)
 
+    def save_subset_as_table(self, filename):
```
Do we want an option to overwrite? Also does this need some path wrangling magic for the "standalone app"? For example, see implementation in this method:

https://github.com/spacetelescope/jdaviz/blob/ee17ed45fcbaa44931112c70c86a7e7201289fcd/jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py#L314

pllim replied:
Actually I see that this plugin already has `_normalize_filename` method that you wrote recently. Should that be used explicitly?

rosteen replied:
`_normalize_filename` is already used before we get to this point, and handling overwrite is also handled in that method.

pllim replied:
Ah OK, hard to tell from the diff. Thanks for the clarification!

pllim reviewed (approved):
Works as advertised but some feedback for possible future work as follows.

When I have multiple spectral regions, I would like the option to save all or some of them in the same file. Currently, I can only save one Subset per file. But what would be nice is something like this in the same file:

```
lower_bound upper_bound
7.030608671135324 7.168932092864794
5.701354428693486 5.876789012350375
```

I have checked that both `astropy.table.QTable` and `specutils.SpectralRegion` could read multi-row ECSV back in.

Even better (and probably only makes sense when we support Subset renaming; also double check how ECSV really store fields with whitespace, this is just pseudocode):

```
name lower_bound upper_bound
"My fav line 1" 7.030608671135324 7.168932092864794
"My other fav line" 5.701354428693486 5.876789012350375
```

pllim commented:
p.s. Not sure if there are any other relevant info you want to also store, like the dataset it was tied to, or the redshift. Again, ideas for future work, not this PR.

rosteen commented:
@pllim Thanks for the comments. This _can_ save out multiple lines to the same file, but it has to be a single subset with multiple subregions rather than multiple spectral subsets.

pllim commented:
Re: https://github.com/spacetelescope/jdaviz/pull/2843#issuecomment-2089194752

Do you think such a distinction should be documented in user doc, maybe at https://jdaviz.readthedocs.io/en/latest/specviz/export_data.html#spectral-regions ? 🤔

kecnry commented on jdaviz/configs/default/plugins/export/export.vue:
```diff
@@ -126,7 +126,7 @@
-      <span class="export-category">Spatial Subsets</span>
+      <span class="export-category">Subsets</span>
       <v-row>
         <span class="category-content v-messages v-messages__message text--secondary">Export spatial subset as astropy region.</span>
```
this description definitely needs generalizing now (while still keeping this information since that was explicitly requested).

kecnry commented on jdaviz/configs/default/plugins/export/export.py:
```diff
@@ -258,6 +259,42 @@ def _is_filename_changed(self, event):
+        elif event['new'] == "ecsv":
+            bad_combo = True
+
+        if bad_combo:
+            raise ValueError(f"Cannot export {self.subset.selected} in "
+                             f"{self.subset_format.selected.upper()} format")
```
This doesn't revert the actual value, and so leaves the plugin in a broken state at which point clicking export or calling `export()` gives an uninformative traceback.  I suspect the code to revert the selection would be messy and perhaps confusing behavior, but maybe we can keep this error message, still show the red text and disable the export button, and then raise that red text as an error on export (all that infrastructure should already exist for the other cases already 🤞 )

To reproduce:
* create a spectral subset and select in the UI
* using the API, request 'fits' as the format.  This error is raised but the value is still set.
* (subsequent attempts to set would not raise an error since the value hasn't changed)
* click or call export

kecnry reviewed (commented):
I was worried about how the automatic changing of format would work wrt the API, but in playing with it, I actually think its ok and quite convenient!  One caveat explained below where the API can still get into trouble, but I think we can find ways around that.

I do think it would then be worth looking into what other places this pattern could be used.  In some cases, I think the red text is still the better choice because it can explain why, but in cases where it is more obvious like this, this really is a nice UX improvement!

Another thing to consider: instead of bumping specutils, we could just have spectral subsets disabled entirely based on the imported version (in this case it probably would make sense to allow selecting the spectral subset but then have the red text say that upgrading specutils is necessary).  I'm ok with the bump though if that keeps things simpler and has other advantages as well.

rosteen commented on jdaviz/configs/default/plugins/export/export.py:
```diff
@@ -258,6 +259,46 @@ def _is_filename_changed(self, event):
+            good_format = [format["label"] for format in self.subset_format_items if
+                           format["disabled"] is False][0]
+            self.subset_format.selected = good_format
+            raise ValueError(f"Cannot export {self.subset.selected} in {event['new']}"
+                             f" format, reverting selection to {self.subset_format.selected}")
```
@kecnry, do you think this error message makes clear enough what is happening?

kecnry replied:
if you want to go the revert route, you could also use `self.subset_format.select_previous()` (which hopefully should be valid still... I can't think of a case where it wouldn't)

kecnry replied:
but the message makes sense to me, let me test it to see if I can find a way to break it and otherwise should be good to approve.

rosteen commented:
@pllim I updated the docs. It actually looks like the export section in `plugins.rst` hadn't been updated since we added more things in addition to viewers, so I added a mention of plugin plots as well.

pllim commented:
Re: docs -- LGTM. Thanks!

rosteen commented on docs/specviz/plugins.rst:
```diff
@@ -320,4 +320,8 @@ using the |icon-line-select| (line selector) tool in the spectrum viewer.
 ======
 
-This plugin allows a given viewer's plot to be exported to various image formats.
+This plugin allows exporting a the contents of a viewer or a plot within a plugin to various image formats.
```
```suggestion
This plugin allows exporting the contents of a viewer or a plot within a plugin to various image formats.
```

rosteen commented on docs/specviz/plugins.rst:
```diff
@@ -320,4 +320,8 @@ using the |icon-line-select| (line selector) tool in the spectrum viewer.
 
-This plugin allows a given viewer's plot to be exported to various image formats.
+This plugin allows exporting the contents of a viewer or a plot within a plugin to various image formats.
+Additionally, spatial and spectral regions can be exported to files, as astropy regions saves as FITS or REG
```
```suggestion
Additionally, spatial and spectral regions can be exported to files, as astropy regions saved as FITS or REG
```

kecnry reviewed (approved):
LGTM now - thanks!
